File: ios/ExpoMapboxNavigationModule.swift
import ExpoModulesCore

public class ExpoMapboxNavigationModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoMapboxNavigation")

    View(ExpoMapboxNavigationView.self) {

      // ── Events (exact parity with Android module Events() list) ─────────────
      Events(
        "onRouteProgressChanged",
        "onRoutesReady",
        "onNavigationFinished",
        "onNavigationCancelled",
        "onRoutesFailed",
        "onArrival",
        "onManeuverBannerPressed"
      )

      // ── Base props ────────────────────────────────────────────────────────────
      Prop("coordinates") { (view: ExpoMapboxNavigationView, v: [[String: Double]]) in
        view.setCoordinates(v)
      }
      Prop("waypointIndices") { (view: ExpoMapboxNavigationView, v: [Int]?) in
        view.setWaypointIndices(v)
      }
      Prop("language") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setLanguage(v)
      }
      Prop("voiceUnits") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setVoiceUnits(v)
      }
      Prop("navigationProfile") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setNavigationProfile(v)
      }
      Prop("excludeTypes") { (view: ExpoMapboxNavigationView, v: [String]?) in
        view.setExcludeTypes(v)
      }
      Prop("mapStyle") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setMapStyle(v)
      }
      Prop("mute") { (view: ExpoMapboxNavigationView, v: Bool) in
        view.setMute(v)
      }
      Prop("maxHeight") { (view: ExpoMapboxNavigationView, v: Double?) in
        view.setMaxHeight(v)
      }
      Prop("maxWidth") { (view: ExpoMapboxNavigationView, v: Double?) in
        view.setMaxWidth(v)
      }
      Prop("useMapMatching") { (view: ExpoMapboxNavigationView, v: Bool) in
        view.setUseMapMatching(v)
      }
      Prop("customRasterTileUrl") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setCustomRasterTileUrl(v)
      }
      Prop("customRasterAboveLayerId") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setCustomRasterAboveLayerId(v)
      }
      Prop("showEndOfRouteFeedback") { (view: ExpoMapboxNavigationView, v: Bool) in
        view.setShowEndOfRouteFeedback(v)
      }
      Prop("showEta") { (view: ExpoMapboxNavigationView, v: Bool) in
        view.setShowEta(v)
      }

      // ── Color customization props (parity with Android) ───────────────────────
      Prop("maneuverBackgroundColorDay") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setManeuverBackgroundColorDay(v)
      }
      Prop("maneuverBackgroundColorNight") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setManeuverBackgroundColorNight(v)
      }
      Prop("maneuverTurnIconColor") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setManeuverTurnIconColor(v)
      }
      Prop("navigationPuckColor") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setNavigationPuckColor(v)
      }
      Prop("navigationPuckImagePath") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setNavigationPuckImagePath(v)
      }
      Prop("navigationPuck3DModelPath") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setNavigationPuck3DModelPath(v)
      }
      Prop("etaBarBackgroundColor") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setEtaBarBackgroundColor(v)
      }
      Prop("etaTextColor") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setEtaTextColor(v)
      }
      Prop("iconButtonColor") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setIconButtonColor(v)
      }
      Prop("iconButtonMutedColor") { (view: ExpoMapboxNavigationView, v: String?) in
        view.setIconButtonMutedColor(v)
      }
    }
  }
}

File: android/src/main/java/expo/modules/mapboxnavigation/ExpoMapboxNavigationModule.kt
package expo.modules.mapboxnavigation

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoMapboxNavigationModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoMapboxNavigation")

        View(ExpoMapboxNavigationView::class) {
            // Events must be declared here — they correspond to EventDispatcher fields in the View
            Events(
                "onRouteProgressChanged",
                "onRoutesReady",
                "onNavigationFinished",
                "onNavigationCancelled",
                "onRoutesFailed",
                "onArrival",
                "onManeuverBannerPressed"
            )

            Prop("coordinates") { view: ExpoMapboxNavigationView, coordinates: List<Map<String, Double>> ->
                view.setCoordinates(coordinates)
            }
            Prop("waypointIndices") { view: ExpoMapboxNavigationView, indices: List<Int>? ->
                view.setWaypointIndices(indices)
            }
            Prop("language") { view: ExpoMapboxNavigationView, language: String? ->
                view.setLanguage(language)
            }
            Prop("voiceUnits") { view: ExpoMapboxNavigationView, units: String? ->
                view.setVoiceUnits(units)
            }
            Prop("navigationProfile") { view: ExpoMapboxNavigationView, profile: String? ->
                view.setNavigationProfile(profile)
            }
            Prop("excludeTypes") { view: ExpoMapboxNavigationView, types: List<String>? ->
                view.setExcludeTypes(types)
            }
            Prop("mapStyle") { view: ExpoMapboxNavigationView, style: String? ->
                view.setMapStyle(style)
            }
            Prop("mute") { view: ExpoMapboxNavigationView, mute: Boolean ->
                view.setMute(mute)
            }
            Prop("maxHeight") { view: ExpoMapboxNavigationView, height: Double? ->
                view.setMaxHeight(height)
            }
            Prop("maxWidth") { view: ExpoMapboxNavigationView, width: Double? ->
                view.setMaxWidth(width)
            }
            Prop("navigationViewportPaddingTop") { view: ExpoMapboxNavigationView, v: Double? ->
                view.setNavigationViewportPaddingTop(v)
            }
            Prop("navigationViewportPaddingLeft") { view: ExpoMapboxNavigationView, v: Double? ->
                view.setNavigationViewportPaddingLeft(v)
            }
            Prop("navigationViewportPaddingBottom") { view: ExpoMapboxNavigationView, v: Double? ->
                view.setNavigationViewportPaddingBottom(v)
            }
            Prop("navigationViewportPaddingRight") { view: ExpoMapboxNavigationView, v: Double? ->
                view.setNavigationViewportPaddingRight(v)
            }
            Prop("useMapMatching") { view: ExpoMapboxNavigationView, use: Boolean ->
                view.setUseMapMatching(use)
            }
            Prop("showEndOfRouteFeedback") { view: ExpoMapboxNavigationView, show: Boolean ->
                view.setShowEndOfRouteFeedback(show)
            }
            Prop("customRasterTileUrl") { view: ExpoMapboxNavigationView, url: String? ->
                view.setCustomRasterTileUrl(url)
            }
            Prop("customRasterAboveLayerId") { view: ExpoMapboxNavigationView, layerId: String? ->
                view.setCustomRasterAboveLayerId(layerId)
            }
            // ── Color customization props ────────────────────────────────────────
            Prop("maneuverBackgroundColorDay") { view: ExpoMapboxNavigationView, color: String? ->
                view.setManeuverBackgroundColorDay(color)
            }
            Prop("maneuverBackgroundColorNight") { view: ExpoMapboxNavigationView, color: String? ->
                view.setManeuverBackgroundColorNight(color)
            }
            Prop("maneuverTurnIconColor") { view: ExpoMapboxNavigationView, color: String? ->
                view.setManeuverTurnIconColor(color)
            }
            Prop("navigationPuckColor") { view: ExpoMapboxNavigationView, color: String? ->
                view.setNavigationPuckColor(color)
            }
            Prop("navigationPuckImagePath") { view: ExpoMapboxNavigationView, path: String? ->
                view.setNavigationPuckImagePath(path)
            }
            Prop("navigationPuck3DModelPath") { view: ExpoMapboxNavigationView, path: String? ->
                view.setNavigationPuck3DModelPath(path)
            }
            Prop("speedLimitPosition") { view: ExpoMapboxNavigationView, position: String? ->
                view.setSpeedLimitPosition(position)
            }
            Prop("showEta") { view: ExpoMapboxNavigationView, show: Boolean? ->
                view.setShowEta(show)
            }
            Prop("etaBarBackgroundColor") { view: ExpoMapboxNavigationView, color: String? ->
                view.setEtaBarBackgroundColor(color)
            }
            Prop("etaTextColor") { view: ExpoMapboxNavigationView, color: String? ->
                view.setEtaTextColor(color)
            }
            Prop("iconButtonColor") { view: ExpoMapboxNavigationView, color: String? ->
                view.setIconButtonColor(color)
            }
            Prop("iconButtonMutedColor") { view: ExpoMapboxNavigationView, color: String? ->
                view.setIconButtonMutedColor(color)
            }
        }
    }
}
